In [2]:
print('HEllo world')

HEllo world


#  importing necessary modules

In [3]:
import pandas as pd
import numpy as np

# 1. Task 1 (Data Preprocessing)


## a)
 Read in the individual CSV-files female coaches.csv, female players.csv,
female teams.csv, male coaches.csv, male players.csv as well as
male teams.csv, and store all the information in stand-alone Pandas-DataFrames
named df_feco (female coaches), df_fepl (female players), df_fete (female teams),
df_maco (male coaches), df_mapl (male players), df_mate (male teams).


In [4]:
# reading  coachs csv file
df_feco=pd.read_csv('female_coaches.csv')
df_maco=pd.read_csv('male_coaches.csv')
print(f'The shape of female_coaches.csv file is {df_feco.shape} and male_coaches.csv file is {df_maco.shape}')
# reading players csv file
df_fepl=pd.read_csv('female_players.csv')
df_mapl=pd.read_csv('male_players.csv')
print(f'The shape of female_players.csv file is {df_fepl.shape} and male_players.csv file is {df_mapl.shape}')
# reading teams
df_fete=pd.read_csv('female_teams.csv')
df_mate=pd.read_csv('male_teams.csv')
print(f'The shape of female_teams.csv file is {df_fete.shape} and male_teams.csv file is {df_mate.shape}')


The shape of female_coaches.csv file is (94, 8) and male_coaches.csv file is (1369, 8)


C:\Users\kashi\AppData\Local\Temp\ipykernel_28956\2433300494.py:7: DtypeWarning: Columns (108) have mixed types. Specify dtype option on import or set low_memory=False.
  df_mapl=pd.read_csv('male_players.csv')


The shape of female_players.csv file is (5035, 109) and male_players.csv file is (180021, 109)
The shape of female_teams.csv file is (231, 54) and male_teams.csv file is (6947, 54)


## b)
Verify the number of total columns for each DataFrame to check whether the numbers
of overall attributes match with the numbers given in the official documentation. For the
players (df_fepl & df_mapl) it should be 109, for the teams (df_fete & df_mate) it
should be 54, and for the coaches (df_feco & df_maco) it should be 8 total attributes.
How many data entries for players, teams, and coaches are provided by the EASD corpus?

In [5]:
# checking the total number of columns in data frame
df_fepl_columns= len(df_fepl.columns)
df_mapl_columns=len(df_mapl.columns)
print(f'There are {df_fepl_columns} columns in female_players.CSV file and {df_mapl_columns} columns in male_players.CSV file.')
df_fete_columns=len(df_fete.columns)
df_mate_columns=len(df_mate.columns)
print(f'There are {df_fete_columns} columns in female_teams.CSV file and {df_mate_columns} columns in male_teams.CSV file.  ')
df_feco_columns=len(df_feco.columns)
df_maco_columns=len(df_maco.columns)
print(f'There are {df_feco_columns} columns in female_feco.CSV file and {df_maco_columns} columns in male_maco.CSV file.  ')

There are 109 columns in female_players.CSV file and 109 columns in male_players.CSV file.
There are 54 columns in female_teams.CSV file and 54 columns in male_teams.CSV file.  
There are 8 columns in female_feco.CSV file and 8 columns in male_maco.CSV file.  


## C)
Add an additional column with the name Sex to each of the 6 DataFrames and fill it
accordingly with the String male or female. Afterwards, combine and merge the two
sex-specific DataFrames for players (df_mapl, df_fepl), teams (df_mate, df_fete)
and coaches (df_maco, df_feco), resulting in a total of 3 new DataFrames, named:
df player, df team, and df coach.

In [6]:
# adding additional columns to all six DataFrames
df_fete['Sex']=np.zeros(len(df_fete))
df_mate['Sex']=np.ones(len(df_mate))

df_mapl['Sex']=np.ones(len(df_mapl))
df_fepl['Sex']=np.zeros(len(df_fepl))

df_maco['Sex']=np.ones(len(df_maco))
df_feco['Sex']=np.zeros(len(df_feco))


In [7]:
# merging the DataFrames with same Sex
players=pd.concat([df_fepl,df_mapl],ignore_index=True)
coaches=pd.concat([df_feco,df_maco],ignore_index=True)
teams=pd.concat([df_fete,df_mate],ignore_index=True)

## d)
Create an updated and reduced version of df player, df team, and df coach as the
very final data foundation, by overwriting the respective DataFrames in such a way that only
the subsequently mentioned attributes (columns) remain:

• Players (27 out of 109): player_id, fifa_version, long_name,
player_positions, value_eur, wage_eur, dob, height_cm, weight_kg,
club_team_id, club_name, league_id, league_name, club_position,
club_jersey_number,club_joined_date, club_contract_valid_until_year,
nationality_id, nationality_name, nation_team_id, preferred_foot,
pace, shooting, passing, dribbling, defending, physic

• Teams (18 out of 54): team_id, fifa_version, team_name, league_id,
league_name, nationality_id, nationality_name, overall, attack,
midfield, defence, coach_id, home_stadium, whole_team_average_age,
captain, penalties, left_corner, right_corner

• Coaches (4 out of 8): coach_id, long_name, dob, nationality_name


In [8]:
# creating df_plyayers with 27 columns.
df_players=players[['player_id', 'fifa_version', 'long_name',
'player_positions', 'value_eur', 'wage_eur', 'dob', 'height_cm', 'weight_kg',
'club_team_id', 'club_name', 'league_id', 'league_name', 'club_position',
'club_jersey_number','club_joined_date', 'club_contract_valid_until_year',
'nationality_id', 'nationality_name', 'nation_team_id', 'preferred_foot',
'pace', 'shooting', 'passing', 'dribbling', 'defending', 'physic']]
# creating df_teams with 18 columns
df_teams=teams[['team_id', 'fifa_version', 'team_name', 'league_id',
'league_name', 'nationality_id', 'nationality_name', 'overall', 'attack',
'midfield', 'defence', 'coach_id', 'home_stadium', 'whole_team_average_age',
'captain', 'penalties', 'left_corner', 'right_corner']]
# cretaing df_coaches with 4 columns
df_coaches=coaches[['coach_id', 'long_name', 'dob', 'nationality_name']]

# checking the shape of every DataFrame
print(f'The shape of the df_players DataFrame is {df_players.shape}')
print(f'The shape of the df_teams DataFrame is {df_teams.shape}')
print(f'The shape of the df_coaches DataFrame is {df_coaches.shape}')

The shape of the df_players DataFrame is (185056, 27)
The shape of the df_teams DataFrame is (7178, 18)
The shape of the df_coaches DataFrame is (1463, 4)


## e)
Check all the individual DataFrames (df_player, df_team, df_coach) for missing values and fill them with appropriate replacements, wherever it is reasonable, in order to guarantee a data collection as clean and error-free as possible. Be careful in which incidents
replacing is not useful, where it is appropriate, and which replacement values make sense
or not. Evaluate, verify, and comment on your decisions, next to reporting a feedback about
the data quality, using various quality criteria.

### I) working on df_coachs dataFrame

In [26]:
# columns in this data set
df_coaches.columns

Index(['coach_id', 'long_name', 'dob', 'nationality_name'], dtype='object')

In [37]:
#checking the missingn values on the data frame
missing_value=df_coaches.isnull().sum().to_frame(name='number of missing values')
df_coaches['dob']=pd.to_datetime(df_coaches['dob'])

C:\Users\kashi\AppData\Local\Temp\ipykernel_28956\3070592389.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_coaches['dob']=pd.to_datetime(df_coaches['dob'])


number of missing values
coach_id                                 0
long_name                                0
dob                                    124
nationality_name                         0

In [45]:
# converting the vlaues into date and time
df_coaches['dob']=pd.to_datetime(df_coaches['dob'])

# creating a column which shows that dob entry of the respective  data sample was missing or not(1=missing,0=not missing)
df_coaches['dob_missing']=df_coaches['dob'].isna().astype(int)

C:\Users\kashi\AppData\Local\Temp\ipykernel_28956\493234047.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_coaches['dob']=pd.to_datetime(df_coaches['dob'])
C:\Users\kashi\AppData\Local\Temp\ipykernel_28956\493234047.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_coaches['dob_missing']=df_coaches['dob'].isna().astype(int)


In [51]:
# checking the mean of the dob values
mean_value=df_coaches['dob'].mean()
df_coaches['dob'].fillna(mean_value,inplace=True)
df_coaches['dob']

C:\Users\kashi\AppData\Local\Temp\ipykernel_28956\3907608065.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_coaches['dob'].fillna(mean_value,inplace=True)


0      1973-02-25 00:00:00.000000000
1      1988-05-29 00:00:00.000000000
2      1976-07-20 00:00:00.000000000
3      1979-07-06 00:00:00.000000000
4      1974-07-21 00:00:00.000000000
                    ...             
1458   1990-10-07 00:00:00.000000000
1459   1967-12-27 00:00:00.000000000
1460   1973-11-03 05:06:29.843166544
1461   1972-11-26 00:00:00.000000000
1462   1974-05-02 00:00:00.000000000
Name: dob, Length: 1463, dtype: datetime64[ns]

# 3. code